README.md
[![NPM version][npm-version-image]][npm-url] [![NPM downloads][npm-downloads-image]][npm-url] [![MIT License][license-image]][license-url] [![Build Status][travis-image]][travis-url]

# Novation-Mobile

Novation-Mobile is a framework using a NodeJS/MongoDB + Mongoose/Socket.IO/Redis stack. It was built by [Novation Mobile](http://www.novationmobile.com) to create scaleable Node.js servers with an emphasis on quick, standardized development.

Get the source from [GitHub](https://github.com/chapinkapa/novation-mobile) or install via NPM

    npm install novation-mobile --save
**Note:** this will take a while. We include all the dependencies to run this.

## Version

0.3.0

## How to use

In a web.js file at your project root, use the following to set up a novation-mobile server:

    var nm = require('novation-mobile');

    var config = {
      appName: 'ExampleApp',
      server: 'Main',
      port: process.argv[2] || 4050,
      useStaticServer: true,
      favicon: 'favicon.ico',
      envLocation: '_env.js',
      preContent: 'routes.js',
      postContent: 'routes2.js',
      mongooseSchemaLocation: '_schema.js',
      viewEngine: 'jade',
      viewDirectory: 'views',
      publicDirectory: 'public',
      servers: ['Main:' + os.hostname()],
      logger:{
        userName: "",
        password: ""
      },
      api: {
          location: "api",
          safeMode: false,
      },
      onlineUsersConfig: {
          roomsToListenTo: ['onlineUsers'],
          expireUser: 20
      }
    };

    nm.extra(__dirname).server(config);

Each option should be customized for your app. 

#### Config Options:

1. **appName**: Name of your app.
1. **server**: Name of the server that the current code is running on.
1. **port:** What port to run server on. Defaults to process.env.PORT and then to 4050.
1. **useStaticServer:** Wether to allow the server to act as a static server for a specified folder. Used with viewEngine, viewDirectory, and publicDirectory. Defaults to true.
1. **favicon:** Location of your favicon. Defaults to 'public'.
1. **[envLocation](#environmental-variables)**: Location of your environmental variables.
1. **[preContent](#routes)**: Location of your routes that run before api routes.
1. **[postContent](#routes)**: Location of your routes that run after api routes.
1. **[mongooseSchemaLocation](#mongoose-schema)**: Location of your mongoose schema. Defaults to '_schema.js'.
1. **viewEngine:** Which view engine to use. Example: jade, html, handlebars, etc.
1. **viewEngine:** Which directory to be used to serve views, if using dynamic views.
1. **publicDirectory:** Which directory to be used as your 'static folder.'
1. **servers**: An array of servers that is used by redis-logger and socket.io-online-users.
1. **logger.username**: username to access the redis-logger
1. **logger.password**: password to access the redis-logger
1. **[api.location](#standard-apis)**: Location of your api folder.
1. **[api.safeMode](#standard-apis)**: Safemode prevents exporting variables other than functions using API.
1. **[api.version](#standard-apis)**: The version number the server should use for internal calls.
1. **onlineUsersConfig:** An object with configuration options to use socket.io-online-users.
1. **onlineUsersConfig.roomsToListenTo**: Socket.io rooms to listen to.
1. **onlineUsersConfig.expireUser**: When a user loses connection, how long to keep the user in memory until deleting them.
1. **ssl**: An object of options to use ssl on your node server.
1. **ssl.key:** Location of key file to use.
1. **ssl.cert:** Location of the cert file to use.
1. **ssl.port:** Port to have your node.js https server run on.
1. **sslRedirect:** Redirect http to https.
1. **dontUseRedisTTL:** do not use a ttl for redis.
1. **ttl:** Time in seconds until redis expires documents. Defaults to 3600.
1. **addSocketsToRoom:** A function that is called every API call that allows you to add a socket/user to a room for socket.io. The function has two arguments: (session, socket);

## Components

### Environmental Variables

##### Location: _env.js

Allows you to set environment variables used throughout the app:

    exports.configureEnvironment = function(app, process) {
      // required variables
      process.env['SESSION_KEY'] = 'my_express.sid';
      process.env['SESSION_SECRET'] = 'exampleSecret';
      process.env['COOKIE_KEY'] = 'ExampleCookie';
      process.env.MONGO_URI = '';
      process.env.REDIS_URI = 'redis://redis:redis@ip:port/dbindex';

      // add your own
      process.env['SOME_API_KEY'] = 'aaa111nnn123';
    };

### Routes

##### Location: routes.js

Allows you to create custom routes for your app.

    exports.content = function(app, io) {
      // you can use this page for additional, custom routes
      app.get('/', function(req, res, next) {
        res.send('This is an example server');
      });
    };

### Standard APIs

##### Location: api/

Allows you to create APIs that can be accessed by both socket.io and by RESTful requests.

Say I want to call the function 'run' under 'SomeAPI'. I can request the API either using ``http://localhost:4050/api/SomeAPI/run`` or by using sockets on the client:

    socket.emit('api', 'SomeAPI', 'run', {
      testData: 'I Am Groot'
    }, function(err, data) {
      if (err) {
        console.log(err);
      } else {
        console.log(data);
      }
    });

The contents of ``api/SomeAPI.js`` then look like:
    exports.run = function() {
      console.log(data.testData); // prints "I Am Groot"

      var number = Math.random();
      if (number < .5) {
        return fn('This is a standard error message.');
      } else {
        return fn(null, {
          data: 'This the standard way to send data back to the client.'
        });
      }
    };
**Note:** extras has the following data:
1. mongoose, access to the mongoose variable.
2. io
3. socket, the particular socket connection, if available
4. connectionType, either socket, http, or internal.
5. fileName, the file that the API is being hit by. 
6. req, if available
7. res, if available
8. method, the method that is being called.
9. IP Address

###### API Middleware Example
    function testSession(data,fn,session,extras,next){
      if(!session){
        return fn("You have to have a session for this.");
      } else {
        return next();
      }
    }

    exports.testSession=API2(testSession,testSession,function(data,fn,session,extras){
      fn(null,"You have a session!");
    });

    exports.fn=function(){
      fn(null,'yay!!');
    };

    exports.staticVariacle=1;

### Mongoose Schema

##### Location: schema.js

Allows you to create a mongoose schema that can be used throughout your app. Configure your file to look like this:

    var mongoose = require('mongoose');
    var Schema = mongoose.Schema;

    exports.User = mongoose.model('User', new Schema({
      createdAt: Date,
      updatedAt: {
        type: Date,
        default: Date.now
      },
      firstName:String,
      lastName:String,
      fullName:String
    }));
**Note:** everything you export in here will be attached to the global scope. It will be accessible throughout your whole server.


[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square
[license-url]: https://github.com/chapinkapa/novation-mobile/blob/master/LICENSE

[npm-version-image]: http://img.shields.io/npm/v/novation-mobile.svg?style=flat-square
[npm-downloads-image]: http://img.shields.io/npm/dm/novation-mobile.svg?style=flat-square
[npm-url]: https://npmjs.org/package/novation-mobile

[travis-image]: http://img.shields.io/travis/chapinkapa/novation-mobile.svg?style=flat-square
[travis-url]: http://travis-ci.org/chapinkapa/novation-mobile

changelog.md
## 0.3.0

Changed the way we handle sessions in our controller. Namely, your APIs can be hit, even if the request does not have a session. To mitigate this, if you are going to use a session, you must make sure that the session exists for every requests. This can be done easily through middleware.

lib/server.js
var clc = require('cli-color');

var clcError = clc.red;
var clcWarn = clc.yellow;

// configure extras and globals
exports.extra = function(projectDir) {
  GLOBAL.projRequire = function(module) {
    if (module == undefined) {
      return;
    }
    if (module.indexOf('/') != 0) {
      module = '/' + module;
    }
    return require(projectDir + module);
  }
  GLOBAL.rootUrl = projectDir;
  return exports;
};

// configure server
exports.server = function() {
  var envLocation, apiLocation, mongooseSchemaLocation, config;
  var fn = arguments[1];
  if (typeof arguments[0] == 'object') {
    config = arguments[0];
    envLocation = config.envLocation;
    //remove in the future. This prevents forcing config object to have apiLocation
    if (!config.api) {
      config.api = {};
    }
    config.api.location = apiLocation = config.api.location ? config.api.location : config.apiLocation;
    mongooseSchemaLocation = config.mongooseSchemaLocation;
  } else {
    throw 'You do not have the right parameters.';
  }
  if (!config) {
    throw 'Must include a config file.';
  }

  config.server = config.server ? config.server : 'Main';
  config.servers = config.servers ? config.servers : [config.server];

  /* Dependencies */
  var express = require('express');
  var mongoose = require('mongoose');
  var cookie = require('cookie');
  var connect = require('connect');
  var session = require('express-session');
  var errorHandler = require('errorhandler');

  /* Vars */
  var app = express();
  var http = require('http');
  http.globalAgent.maxSockets = Infinity;
  var server = http.createServer(app);
  var RedisStore = require('connect-redis')(session);
  var os = require('os');
  var fs = require('fs');
  var https = require('https');
  var url = require('url');
  var signature = require('cookie-signature');

  if (typeof config.ssl == 'object') {
    config.ssl.key = config.ssl.key != undefined ? config.ssl.key : 'ssl/key.pem';
    config.ssl.cert = config.ssl.cert != undefined ? config.ssl.cert : 'ssl/cert.pem';
    var credentials = {
      key: fs.readFileSync(rootUrl + '/' + config.ssl.key),
      cert: fs.readFileSync(rootUrl + '/' + config.ssl.cert)
    };
    config.serverSSL = https.createServer(credentials, app);
  }

  projRequire(envLocation).configureEnvironment(app, process);
  if (!process.env.REDIS_URI) {
    throw 'Must define "REDIS_URI" in the environment variables.';
  } else if (!process.env.SESSION_KEY) {
    throw 'Must define "SESSION_KEY" in the environment variables.';
  } else if (!process.env.SESSION_SECRET) {
    throw 'Must define "SESSION_SECRET" in the environment variables.';
  } else if (!process.env.COOKIE_KEY) {
    throw 'Must define "COOKIE_KEY" in the environment variables.';
  } else if (!process.env.MONGO_URI) {
    throw 'Must define "MONGO_URI" in the environment variables.';
  }

  var logger;
  if (config.logger) {
    var servers;
    if (config.servers) {
      servers = config.servers;
    }
    logger = require('redis-logger').init(servers, os.hostname());
  }

  var redisUrl = url.parse(process.env.REDIS_URI);
  app.set('redisHost', redisUrl.hostname);
  app.set('redisPort', redisUrl.port);

  if (redisUrl.auth) {
    var redisAuth = redisUrl.auth.split(':');
    app.set('redisDb', redisAuth[0]);
    app.set('redisPass', redisAuth[1]);
  }

  if (redisUrl.path) {
    app.set('redisDbIndex', redisUrl.path.slice(1, redisUrl.path.length));
  } else {
    app.set('redisDbIndex', 0);
  }

  var cookieParser = require('cookie-parser')(process.env.COOKIE_KEY);
  app.use(cookieParser);
  app.use(function(req, res, next) {
    res.header('Access-Control-Allow-Credentials', true);
    res.header('Access-Control-Allow-Origin', "*");
    res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS');
    res.header('Access-Control-Allow-Headers', 'X-Requested-With, X-HTTP-Method-Override, Content-Type, Accept');
    next();
  });

  var redisStoreConfig = {
    url: process.env.REDIS_URI
  };
  if (!config.dontUseRedisTTL) {
    redisStoreConfig.ttl = config.ttl || 3600;
  }
  var sessionStore = new RedisStore(redisStoreConfig);

  /* Redis Setup */
  app.use(session({
    key: process.env.SESSION_KEY,
    secret: process.env.SESSION_SECRET,
    store: sessionStore
  }));

  if (config.sessionExpireDate != undefined) {
    app.use(function(req, res, next) {
      req.session.cookie.expires = config.sessionExpireDate;
      next();
    })
  }
  if (app.get('env') === 'development') {
    app.use(errorHandler({
      dumpExceptions: true,
      showStack: true
    }));
  } else {
    process.addListener('uncaughtException', function(err) {
      console.error((new Date).toUTCString() + ' uncaughtException:', err.message);
      console.error(err.stack);
      process.exit();
    });
  }
  // for file uploads
  var bodyParser = require('body-parser');

  app.use(bodyParser.urlencoded({
    extended: true,
    limit: '2048mb'
  }));

  app.use(bodyParser.json({
    limit: '2048mb'
  }));

  // favicon
  var favicon = rootUrl + '/public/img/favicon.ico'
  if (config.favicon != undefined) {
    favicon = rootUrl + '/' + config.favicon;
  }

  app.use(require('serve-favicon')(favicon), function() {
    console.log(err, favicon_url)
  });

  app.use(require('method-override')());
  app.use(errorHandler());
  if (app.get('env') == 'production') {
    //This is ssl, add when ready
    app.all('*', function(req, res, next) {
      if (req.url.indexOf('healthcheck') !== -1) {
        res.send('server is running');
        return;
      }
      if (config.sslRedirect) {
        if (req.headers.host == undefined) {
          res.send('');
        } else if (req.headers.host.match(/^www/) !== null) {
          res.redirect('https://' + req.headers.host.replace(/^www\./, '') + req.url);
        } else if (req.protocol.indexOf('https') === -1) {
          res.redirect('https://' + req.headers.host + req.url);
        } else {
          next();
        }
      } else {
        next();
      }
    });
  }

  // assuming io is the Socket.IO server object!
  var io = require('socket.io')(server);
  io.use(function(socket, next) {
    var handshakeData = socket.request;
    if (handshakeData.headers.cookie) {
      handshakeData.cookie = cookie.parse(handshakeData.headers.cookie);
      try {
        var real_sid = handshakeData.cookie[process.env.SESSION_KEY].replace('s:', '');
        real_sid = signature.unsign(real_sid, process.env.SESSION_SECRET);

        handshakeData.sessionID = real_sid;
      } catch (err) {
        console.log(clcError(err));
        return next('Looks like we\'ve got a hacker on our hands.', false);
      }

      if (handshakeData.cookie[process.env.SESSION_KEY] == handshakeData.sessionID) {
        console.log(clcError('Cookie is invalid'));
        return next('Cookie is invalid.', false);
      }

    } else {
      return next('No cookie transmitted.', false);
    }
    next(null, true);
  });

  var ioRedis = require('socket.io-redis');
  var redisIO = require('redis');
  var pubIO = redisIO.createClient(app.set('redisPort'), app.set('redisHost'), {
    return_buffers: true
  });
  var subIO = redisIO.createClient(app.set('redisPort'), app.set('redisHost'), {
    return_buffers: true
  });

  pubIO.auth(app.set('redisPass'), function(err) {
    if (err) {
      throw err;
    }
  });
  subIO.auth(app.set('redisPass'), function(err) {
    if (err) {
      throw err;
    }
  });

  // for redis online users
  io.adapter(ioRedis({
    key: 'io:' + config.appName,
    pubClient: pubIO,
    subClient: subIO
  }));

  if (config.onlineUsersConfig) {
    var client = redisIO.createClient(app.set('redisPort'), app.set('redisHost'), {});
    client.select(app.set('redisDbIndex') || 0, function(err) {
      if (err) {
        throw err;
      }
    });
    client.auth(app.set('redisPass'), function(err) {
      if (err) {
        throw err;
      }
    });
    config.onlineUsersConfig.io = io;
    config.onlineUsersConfig.dataClient = client;
    config.onlineUsersConfig.appName = config.appName;
    config.onlineUsersConfig.serverName = config.server + ':' + os.hostname();
    var OnlineUsers = require('socket.io-online-users');
    io.clients = new OnlineUsers(config.onlineUsersConfig);
  }

  if (!config.turnOffAwesomeLogs && logger) {
    logger.logPage(app, config.logger.username, config.logger.password, '/redis-logger.html').socket(io);
  }
  mongoose.connect(process.env.MONGO_URI, function(err) {
    if (err) {
      return console.log(clcError('Connecting to mongoose error::', err));
    }
  });

  mongoose.schema = projRequire(mongooseSchemaLocation);
  require('./helpers.js').makeGlobal(config, mongoose, io).attachAPI();
  if (config.preContent) {
    projRequire(config.preContent).content(app, io, mongoose);
  }

  require('./controller.js').sockets(app, io, sessionStore, mongoose, apiLocation);

  if (config.postContent) {
    projRequire(config.postContent).content(app, io, mongoose);
  }
  if (config.useStaticServer != undefined) {
    if (config.viewEngine) {
      app.set('view engine', config.viewEngine);
    } else {
      app.set('view engine', 'jade');
    }
    if (config.viewDirectory) {
      app.set('views', rootUrl + '/' + config.viewDirectory);
    } else {
      app.set('views', rootUrl + '/views');
    }
    app.locals.basedir = app.get('views');
    if (config.publicDirectory) {
      app.use(express.static(rootUrl + '/' + config.publicDirectory));
    } else {
      app.use(express.static(rootUrl + '/public'));
    }
  }

  var port = process.env.PORT || config.port || 4050;
  server.listen(port);
  console.log('NM server started on port ' + port);
  if (fn) {
    return fn();
  }

  if (typeof config.ssl == 'object') {
    var sslPort;
    if (process.env.PORT) {
      sslPort = 443;
    } else if (config.ssl.port) {
      sslPort = config.ssl.port
    } else {
      sslPort = 5051;
    }
    config.serverSSL.listen(sslPort);
    console.log('NM server started on ssl port ' + sslPort);
    if (fn) {
      return fn();
    }
  }
};
